File: 5-longest-palindromic-substring/longest-palindromic-substring.cpp
class Solution {
public:
    string longestPalindrome(string s) {
        // dp[i][j] tells us if substring from i to j is a palindrome
        int n = s.size();
        vector<vector<bool>> dp(n, vector<bool>(n, false));
        int longest = 1;
        int startIndex = 0;
        int endIndex = 0;

        for (int i = 0; i < n; i++) {
            // every single character is a palindrome by itself
            dp[i][i] = true;

            for (int j = 0; j < i; j++) {
                if (s[j] == s[i] && (i-j <= 2 || dp[j+1][i-1])) {
                    dp[j][i] = true;
                    if (i - j + 1 > longest) {
                        longest = i - j + 1;
                        startIndex = j;
                        endIndex = i;
                    }
                }
            }
        }

        return s.substr(startIndex, endIndex-startIndex+1);
    }
};

File: 5-longest-palindromic-substring/longest-palindromic-substring.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        n = len(s)
        longest = 0
        startIndex = 0
        endIndex = 0

        # dp[i][j] means substring from i to j is a palindrome
        dp = [n * [0] for _ in range(n)]

        for i in range(n):
            dp[i][i] = 1 # every letter itself is a palindrome
            for j in range(n):
                # same letter AND (only one letter in middle or middle is palindrome)
                if s[j] == s[i] and (i-j <= 2 or dp[j+1][i-1]):
                    dp[j][i] = 1
                    if (longest < i-j):
                        longest = i - j
                        startIndex = j
                        endIndex = i
        
        return s[startIndex:endIndex + 1]